rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+```
+
+**TODO: should this be a choose one or a produce complete ordering? The `alreadyFailed` argument is intended to capture previous choices that failed, to remove them from the pool, on the theory choosing 1 is probably easier and more efficient**
+
+Such a function MUST be deterministic from the input nodes and preferences. Eventually we imagine a space having a configured CID that represents a custom implementation of such a function.
+
+Let's say the chosen storage node is:
+
+```json
+{
+  "did": "did:web:node1.storage",
+  "w3upEndpoint": "https://node1.storage"
+}
+```
+
+3. The service invokes `blob/allocate` remotely on the chosen node:
+
+```json
+{ // "/": "bafy..add"
+  "cmd": "/blob/allocate",
+  "aud": "did:web:node1.storage",
+  "iss": "did:web:web3.storage",
+  "sub": "did:web:node1.storage",
+  "args": {
+    "blob": {
+      // multihash of the blob as byte array
+      "digest": { "/": { "bytes": "mEi...sfKg" } },
+      // size of the blob in bytes
+      "size": 2_097_152,
+    },
+    "storagePreferences": {
+    }
+  }
+}
+```
+
+Legacy invocation:
+
+```json
+{ // "/": "bafy..add"
+  "aud": "did:web:node1.storage",
+  "iss": "did:web:web3.storage",
+  "att": [
+    {
+      "can": "/web3.storage/blob/allocate",
+      "nb": "did:web:node1.storage",
+      "with": {
+        "blob": {
+          // multihash of the blob as byte array
+          "digest": { "/": { "bytes": "mEi...sfKg" } },
+          // size of the blob in bytes
+          "size": 2_097_152,
+        },
+        "storagePreferences": {
+        }
+      }
+    }
+  ]
+}
+```
+
+Receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..allocate" },
+      // refers to the receipt corresponding to the above invocation
+      { 
+        "out": {
+          "ok": {
+            // result of the add is the content (location) commitment
+            // that is produced as result of "bafy..accept"
+            "size": 2_097_152,
+            "address": {
+              "url": "https://node1.storage/upload/xxxa-presigned",    
+              "headers": {},
+              "expires": 1712903125
+            }
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafyreia5tctxekbm5bmuf6tsvragyvjdiiceg5q6wghfjiqczcuevmdqcu"
+  },
+  "out": {
+    "ok": {
+      "size": 2_097_152,
+      "address": {
+        "url": "https://node1.storage/upload/xxxa-presigned",    
+        "headers": {},
+        "expires": 1712903125
+      }
+    }
+  },
+  "fx": [],
+}
+```
+
+4. On success, router creates a success receipt for the original user
```

Comment:
I'm not sure we need this, it does not matter who did the allocation, we just want user to upload there and issue their own receipt when they're done. If there is reason I'm missing, it's probably best to add a segment elaborating it.

Reply:
I would also expect this state to be added into the `BlobAllocations` state.

Reply:
blob allocate is included